# Analysis of integration site distributions and relative clonal abundance for subject `r sanitize(patient)`
`r format(Sys.Date(), "%b %d %Y")`

```{r setup,echo=FALSE}
opts_chunk$set(fig.path='figureByPatient/', fig.align='left', comment="",
               echo=FALSE, warning=FALSE, error=TRUE, message=FALSE)
options(knitr.table.format = 'html')
```

## Introduction

The attached report describes results of analysis of integration site distributions and relative abundance for samples from gene therapy trials. For cases of gene correction in hematopoietic stem cells, it is possible to harvest blood cells and analyze the distributions of integration sites. Frequency of isolation information can provide information on the clonal structure of the population. This report summarizes results for subject `r sanitize(patient)` over time points `r sanitize(paste(timepoint,collapse=", "))` in UCSC genome draft `r sanitize(freeze)`. 

The samples studied in this report, the numbers of sequence reads, and unique integration sites available for this subject are shown below. 

```{r summaryTable,results="asis"}
cols <- c("posID","clonecount","Replicate","estAbundance1",
          "Trial","GTSP","patient","timepoint","celltype","enzyme","VCN")
toprint <- sites.qc[,cols]
summaryTable <- ddply(toprint, .(Trial, GTSP, patient, timepoint, celltype), 
                      function(x) {
                        data.frame(TotalReads=sum(x$clonecount), 
                                   UniqueSites=length(unique(x$posID)),
                                   Replicates=length(unique(x$Replicate)),
                                   Enzymes=paste(sort(unique(x$enzyme)),
                                                 collapse=','),
                                   VCN=paste(sort(unique(x$VCN)),collapse=','),
                                   S.chao1=getPopEstimates(
                                     acast(x, posID~Replicate, fun.aggregate=sum, 
                                           value.var="estAbundance1", fill=0)))
                      })
tps <- sortTimePoints(summaryTable$timepoint)  
summaryTable$timepoint <- factor(summaryTable$timepoint, levels=names(tps))
summaryTable <- arrange(summaryTable,patient,timepoint,celltype)                

kable(summaryTable[,setdiff(names(summaryTable),c('Replicates','S.chao1'))], 
      caption="Sample Summary Table", row.names=FALSE, format="html", digits = 0)
```

## Population Size

Under most circumstances only a subset of sites will be sampled. We thus include an estimate of sample size based on frequency of isolation information from the SonicLength method [(Berry, 2012)](http://www.ncbi.nlm.nih.gov/pubmed/22238265). 
The 'S.chao1' column denotes the estimated population size derived using Chao estimate [(Chao, 1987)](http://www.ncbi.nlm.nih.gov/pubmed/3427163). If sample replicates were present then estimates were subjected to jackknife bias correction.

We also quantify population clone structure using Gini Coefficients. The Gini coefficient provides a measure of inequality in clonal abundance in each sample. The coefficient equals zero when all sites are equally abundant (polyclonal) and increases as fewer sites account for more of the total (oligoclonal). 

The table below summarizes sample population for each timepoint & celltype combination. 

```{r ChaoGiniTable,results="asis"}
# calculate unique integration sites per timepoint + celltype along with the gini index to be displayed on the barplots #
# gini index should be calculated by amount of disproportionality in estAbundance1Prop #
test2 <- unique(sites.qc[,c("Position", "Aliasposid", "estAbundance1Prop", "GTSP",
                            "timepoint", "celltype", "patient")])

sites.props <- ddply(test2, .(patient,timepoint,celltype,GTSP), 
                     summarize, 
                     UniqueSites=length(unique(Aliasposid)), 
                     Gini=gini(estAbundance1Prop))
sites.props$Gini <- round(sites.props$Gini,digits=2)

summaryTable <- merge(summaryTable, sites.props)
tps <- sortTimePoints(summaryTable$timepoint)
summaryTable$timepoint <- factor(summaryTable$timepoint, levels=names(tps))
summaryTable <- arrange(summaryTable, patient,timepoint,celltype)  

kable(summaryTable[,setdiff(names(summaryTable),c('Trial','GTSP','TotalReads'))], 
      caption="Sample Population Summary", format="html", digits = 2, 
      row.names=FALSE)
```

The graph below visualizes population based summaries at a function of time.

```{r pop_graphs, fig.width=8, fig.height=6}
## get new gini by timepoint only ##
res <- unique(sites.qc[,c("posID", "timepoint", "estAbundance1")])
pops <- ddply(res, .(timepoint), summarize,
              UniqueSites=length(unique(posID)),
              Gini=gini(getPropsAndRanks(estAbundance1, posID, "A", "")$Prop))

## get Shannon diversity by timepoint only ##
H <- diversity(acast(res, timepoint~posID, fill=0, value.var="estAbundance1", 
                     fun.aggregate=sum))
pops$Shannon <- H[pops$timepoint]

## get new chao by timepoint only ##
res <- unique(sites.qc[,c("posID", "timepoint", "Replicate", "estAbundance1")])
res <- split(res, res$timepoint)
H <- sapply(res, function(x) 
  getPopEstimates(acast(x, formula=posID~Replicate, fun.aggregate=sum, 
                        value.var="estAbundance1", fill=0)))
names(H) <- names(res)
pops$S.Chao <- H[pops$timepoint]
rm(res)

tps <- sortTimePoints(pops$timepoint)
pops$timepoint <- factor(pops$timepoint, levels=names(tps))

pU <- qplot(data=pops, timepoint, UniqueSites, geom="bar", stat='identity',
            ylab="Unique Integration Sites")
pG <- qplot(data=pops, timepoint, Gini, geom="bar", stat='identity',
            ylab="Gini Index")
pC <- qplot(data=pops, timepoint, S.Chao, geom="bar", stat='identity',
            ylab="Estimated Population (S.Chao)")
ymax <- max(pops$Shannon)+.5
pS <- qplot(data=pops, timepoint, Shannon, geom="bar", stat='identity',
            ylab="Shannon Index") +
  scale_y_continuous(breaks=seq(0, ymax, by=2),limits=c(0, ymax))

grid.arrange(pU, pG, pC, pS, ncol=2)

```

## Relative abundance of cell clones

The relative abundance of cell clones is summarized in the attached stacked bar graphs.  The cell fraction studied is named at the top, the time points are marked at the bottom. The different bars in each panel show the major cell clones, as marked by integration sites.  A key to the sites is shown at the right.  Each integration site is named by the nearest gene. The '*' indicates if the integration site is with the transcription unit for that gene, the '~' indicates a cancer related gene, and the '!' indicates a gene of interest from previous gene therapy trials (these include genes involved in adverse events, and genes at clustered integration sites from the first SCID trial).

Integration sites were recovered using ligation mediated PCR after random fragmentation of genomic DNA, which reduces recovery biases compared with restriction enzyme cleavage. Relative abundance was not measured from read counts, which are known to be inaccurate, but from marks introduced into DNA specimens prior to PCR amplification using the SonicLength method [PMID:22238265](http://www.ncbi.nlm.nih.gov/pubmed/22238265).

*Note*: In the barplots below any sites with Estimated Relative Abundance below `r percent(abundCutoff)` are binned as *LowAbund*. In cases where the legends are getting clipped for a certain celltype;  it is safe to assume that at a given timepoint numerous integration sites were equally abundant hence scoring the same rank.

```{r barPlots_prep}
# generalize low abundant gene names #
sites.qc$siteType <- with(sites.qc,
                          ifelse(estAbundance1Prop>=abundCutoff,geneType, "LowAbund"))

sums <- aggregateSiteTypes(sites.qc)
sums$timepoint <- sub("(.+):(.+)","\\1",sums$Alias)
sums$celltype <- sub("(.+):(.+)","\\2",sums$Alias)
patternSearch <- "barPlots"
```

```{r barPlots, fig.height=5, fig.show='hide'}
### Stacked Barplot: tp vs geneType ###
for(cell in celltype) {
  toplot <- droplevels(subset(sums,celltype==cell))
  if(nrow(toplot)>0) {
    # reorder siteType2 by their abundance #
    toplot$siteType2 <- with(toplot,
                             factor(siteType,
                                    levels=unique(siteType[order(Ranks)]),
                                    ordered=TRUE))            
    toplot <- arrange(toplot, timepoint, Props)
    
    # reorder timepoints by days #
    tps <- sortTimePoints(toplot$timepoint)
    toplot$timepoint <- factor(toplot$timepoint, levels=names(tps))                                                                               
    
    # set custom color scale #
    siteTypeCols <- structure(gg_color_hue(length(levels(toplot$siteType2))),
                              names=levels(toplot$siteType2))
    siteTypeCols["LowAbund"] <- "#E0E0E0"
    
    p <- ggplot(data=toplot, aes(timepoint, Props, fill=siteType2)) + 
      geom_bar(stat="identity") + facet_wrap(~celltype,scales="free_x") + 
      scale_y_continuous(labels=percent, expand=c(0,0)) + 
      scale_fill_manual(name="SiteType", values=siteTypeCols) +
      guides(fill=guide_legend(ncol=ifelse(length(siteTypeCols)>20,2,1))) +
      labs(y="Estimated Relative Abundance by Pat-Tp-Cell", x="Timepoint") + 
      theme(legend.title=element_blank(), 
            axis.text.x=element_text(angle=45,hjust=1,vjust=1))
    print(p)  
  }
}
```
```{r barPlots_Table, results='asis'}
plotFiles <- list.files(path="figureByPatient", 
                        pattern=paste0(patternSearch,".*.png"), full.names=TRUE)
plotFiles <- paste0("<img src='",plotFiles,"' alt='barPlots'>")
plotFiles <- matrix(plotFiles, ncol=2)
dups <- names(which(table(plotFiles)>1))
if(length(dups)>0) {
  plotFiles[tail(which(plotFiles==dups),1)] <- ""
}
kable(plotFiles, row.names=FALSE,format="html")
```

Here is another way to perceive top ranking integration sites by genes within each celltype.

```{r sitetype_heatmap, fig.width=12, fig.height=11}
sums$siteType2 <- sums$siteType
counts <- sort(table(sums$siteType2))
sums$siteType2 <- factor(sums$siteType2, levels=names(counts), ordered=TRUE)  
if(length(levels(sums$siteType2))<50) {
  sums <- aggregateSiteTypes(sites.qc, siteTypeVar="geneType")
  abundCutoff2 <- abundCutoff - 0.01 
  sums$siteType2 <- with(sums, ifelse(Props>=abundCutoff2, siteType, "LowAbund"))
  sums$siteType2 <- with(sums, 
                         ifelse(siteType %in% siteType2[siteType2!="LowAbund"],
                                      siteType, siteType2))  
  counts <- sort(table(sums$siteType2))
  sums$siteType2 <- factor(sums$siteType2, levels=names(counts), ordered=TRUE)  
  
  while(length(levels(sums$siteType2))<50) {
    abundCutoff2 <- abundCutoff2 - 0.01
    sums$siteType2 <- with(sums, ifelse(Props>=abundCutoff2, siteType, "LowAbund"))
    sums$siteType2 <- with(sums, 
                           ifelse(siteType %in% siteType2[siteType2!="LowAbund"],
                                        siteType, siteType2))  
    counts <- sort(table(sums$siteType2))
    sums$siteType2 <- factor(sums$siteType2, levels=names(counts), ordered=TRUE)  
  }
 
  cat("Since using", percent(abundCutoff), "abundance cut off yeilds only small number of sites, we lower it down to", percent(abundCutoff2), "so it can capture a larger slice of the data.")
  
  sums <- ddply(.data=sums, .(Alias,siteType2), summarize, Props=sum(Props))  
  sums$timepoint <- sub("(.+):(.+)","\\1",sums$Alias)
  sums$celltype <- sub("(.+):(.+)","\\2",sums$Alias)
}

tps <- sortTimePoints(sums$timepoint)
sums$timepoint <- factor(sums$timepoint, levels=names(tps))
counts <- count(sums, c("timepoint","celltype"))
sums$celltype <- factor(sums$celltype, levels=unique(counts$celltype))

p <- qplot(data=sums, timepoint, siteType2, fill=Props, 
           geom='tile', xlab="Timepoint", ylab="SiteType") +
  scale_fill_continuous(name='Relative\nAbundance', labels=percent,
                        low="#E5F5E0", high="#2B8CBE") + 
  facet_grid(.~celltype, scales="free_x", space="free_x") +
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1))
print(p)
```

### Relative abundance of cell clones using only read counts
```{r barPlotsClonecount_prep}
# generalize low abundant gene names #
sites.qc$siteType <- with(sites.qc,
                          ifelse(ccProp>=abundCutoff, geneType, "LowAbund"))

sums <- aggregateSiteTypes(sites.qc,valueVar="clonecount")
sums$timepoint <- sub("(.+):(.+)","\\1",sums$Alias)
sums$celltype <- sub("(.+):(.+)","\\2",sums$Alias)
patternSearch <- "barPlotsClonecounts"
```

```{r barPlotsClonecounts, ref.label='barPlots', fig.height=5, fig.show='hide'}
```
```{r barPlotsClonecounts_table, ref.label='barPlots_Table', results='asis'}
```

## Sites shared between different cell types: overlap matrix

A question of interest in stem cell gene therapy centers on the question of which types of cells share common progenitor cells, as indicated by common integration sites.  Below we first summarize overlap in a matrix.  The numbers of integration sites are shown for each sample along the diagnal.  The overlaps among all pairs are shown as the values in the matrix.  Note that this is based on presence-absence information, and not scaled by abundance.  In interpreting this data, care must be taken to account for possible overlap due to impurities in cell fractionation or cross-overs between samples during integration site recovery.

```{r overlapMatrix,results="asis"}
test2 <- unique(sites.qc[,c("posID","estAbundance1Prop","timepoint",
                            "celltype","patient","Trial")])

if(any(rowSums(pmin(xtabs(~timepoint+celltype,test2),1))>1)) {
  for(tp in names(timepoint)) {
    toplot <- droplevels(subset(test2,timepoint==tp))
    if(nrow(toplot)>0 & length(unique(toplot$celltype))>1) {
      dat <- with(toplot,split(posID,celltype))
      res <- sapply(names(dat), function(y) 
        sapply(names(dat), function(z) length(intersect(dat[[y]],dat[[z]])) ) )
      tabcap <- paste("Patient:",patient,Trial,
                      ", Tp:",tp,", UniqueSites:",length(unique(toplot$posID)))          
      kable(res, format="html", caption=tabcap)
      }
    }    
} else {
  cat("<strong>Not enough celltypes available per timepoint to perform this analysis.</strong>")
}
```

## Shared sites between different cell types: Jaccard Distance

A common means of summarizing shared membership is by using the Jaccard Distance. The Jaccard Distance is calculated from shared membership information using the following equation.

![Jaccard](jaccard_equation.png)

A heat map summarizing the Jaccard Distances among all pairs of samples is shown below. Perfect shared membership has value 0, no shared membership has value 1.

```{r JaccardHeatmap, fig.height=5, fig.show='hide'}
### Jaccard Index heatmap ###
test2 <- unique(sites.qc[,c("posID","estAbundance1Prop","timepoint",
                            "celltype","patient","Trial")])

if(any(rowSums(pmin(xtabs(~timepoint+celltype,test2),1))>1)) {
  for(tp in names(timepoint)) {
    toplot <- droplevels(subset(test2, timepoint==tp))
    if(nrow(toplot)>0 & length(unique(toplot$celltype))>1) {
      plot.frame <- cast(toplot, celltype ~ posID, 
                         value="estAbundance1Prop",fill=0)                
      res <- vegdist(plot.frame,"jaccard",upper=T,diag=T)
      
      pheatmap(as.matrix(res), 
               main=paste("Patient:",patient,Trial,", Tp:",tp))
    }
  }
}
```

```{r jHeatmap_Table, results='asis'}
plotFiles <- list.files(path="figureByPatient", pattern="JaccardHeatmap.*.png", 
                        full.names=TRUE)
if(length(plotFiles)>0) {
  plotFiles <- paste0("<img src='",plotFiles,"' alt='JaccardHeatmaps'>")
  plotFiles <- matrix(plotFiles, ncol=2)
  dups <- names(which(table(plotFiles)>1))
  if(length(dups)>0) {
    plotFiles[tail(which(plotFiles==dups),1)] <- ""
    }
  kable(plotFiles, row.names=FALSE,format="html")
} else {
  cat("<strong>Not enough celltypes available per timepoint to perform this analysis.</strong>")
}
```

## Longitudinal behavior of major clones

When multiple time points are available, it is of interest to track the behavior of the most abundant clones.  A plot of the relative abundances of major clones, based on output from SonicLength, is shown below. For cases where only a single time point is available, the data is just plotted as unlinked points. 

```{r ParallelLines, fig.width=10, fig.height=10}
### Parallel lines plot ###
toplot <- unique(sites.qc[,c("posID","estAbundance1Prop","estAbundance1PropRank",
                             "siteType","geneType","timepoint","celltype",
                             "patient","Trial")])
if(nrow(toplot)>0) {
  tps <- sortTimePoints(toplot$timepoint)
  toplot$timepoint <- factor(toplot$timepoint, levels=names(tps))
  toplot$SiteRank <- factor(with(toplot,
                                 ifelse(estAbundance1PropRank<4,
                                        estAbundance1PropRank,">3")),
                            levels=c(1:3,">3"))
  
  if(length(unique(toplot$celltype))>8) {
    toplot$celltype <- abbreviate(toplot$celltype)
  }
  
  print(ggplot(toplot,aes(x=timepoint, y=estAbundance1Prop)) + 
          geom_point(size=.5) + 
          geom_line(aes(colour=posID, group=posID), alpha=.5, show_guide=FALSE) + 
          facet_wrap(~celltype, scales="free") + 
          ggtitle(paste("Patient:",patient,Trial)) + xlab("Timepoint") + 
          scale_y_continuous(name="Estimated Relative Abundance by Pat-Tp-Cell",
                             labels=percent,expand=c(0,0)) +
          theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1)))
}
```

### Longitudinal behavior of major clones by read counts
```{r ParallelLinesClonecount, fig.width=10, fig.height=10}
### Parallel lines plot ###
toplot <- unique(sites.qc[,c("posID","ccProp","ccPropRank",
                             "siteType","geneType","timepoint","celltype",
                             "patient","Trial")])
if(nrow(toplot)>0) {
  tps <- sortTimePoints(toplot$timepoint)
  toplot$timepoint <- factor(toplot$timepoint, levels=names(tps))
  toplot$SiteRank <- factor(with(toplot,
                                 ifelse(ccPropRank<4,
                                        ccPropRank,">3")),
                            levels=c(1:3,">3"))
  
  if(length(unique(toplot$celltype))>8) {
    toplot$celltype <- abbreviate(toplot$celltype)
  }
  
  print(ggplot(toplot,aes(x=timepoint, y=ccProp)) + 
          geom_point(size=.5) + 
          geom_line(aes(colour=posID, group=posID), alpha=.5, show_guide=FALSE) + 
          facet_wrap(~celltype, scales="free") + 
          ggtitle(paste("Patient:",patient,Trial)) + xlab("Timepoint") + 
          scale_y_continuous(name="Read Count Abundance by Pat-Tp-Cell",
                             labels=percent,expand=c(0,0)) +
          theme(axis.text.x=element_text(angle=45,hjust=1,vjust=1)))
}
```

## Integration sites near LMO2

Integration sites near the LMO2 gene are of special interest, because integration sites near these genes have now been associated with multiple adverse events after hematopoietic stem cell gene therapy. For this reason, we have cataloged all integration sites for which LMO2 is the nearest cancer-related gene over the specimens for all subjects. Results are summarized below.

Below is a table of sites for which LMO2 is the closest proto-onconogene. Note that many might be far from the integration site (DistTo5p column). Column headings are defined below. The table is also shown as a scatter plot where the y-axis shows abundance of sites and x-axis is distance to the nearest proto-onconogene 5' end. 
```{r LMO2}
geneOfInterest <- "LMO2"
printIt <- TRUE
```

```{r geneOfInterest, results="asis", fig.width=10, fig.height=7}
### geneOfInterest sites ###
rows <- with(sites.qc, grepl(geneOfInterest,nrstOncoGeneName) | 
               grepl(geneOfInterest,nrst5pOncoGeneName))

if(any(rows)) {
  ### geneOfInterest barplot ###                         
  toplot <- sites.qc[rows,c("timepoint","nrst5pOncoGeneDist",
                            "celltype","estAbundance1Prop")]
  tps <- sortTimePoints(toplot$timepoint)
  toplot$timepoint <- factor(toplot$timepoint, levels=names(tps))
  print(qplot(data=toplot, x=nrst5pOncoGeneDist, y=estAbundance1Prop, 
              colour=timepoint, geom="point", size=3, alpha=.95) + 
          scale_x_continuous(name=paste("Distance to",geneOfInterest,"5' end (bp)"),
                             labels=comma) + 
          scale_y_continuous(name="Estimated Relative Abundance by Pat-Tp-Cell",
                             labels=percent, limits=c(0,1)) +
          facet_wrap(~celltype) + guides(alpha=FALSE, size=FALSE) +
          theme(axis.text.x=element_text(angle=45,hjust=1)))

  ### geneOfInterest sites table description ###                         
  if(printIt) {
    res <- c("Trial"="Shorthand name for clinical trial", 
             "Pat"="Patient ID", 
             "Tp"="Time point", 
             "Cell"="Cell type", 
             "posID"="Chromosomal location of the integration site; first chromosome number, then coordinate. The + or - indications orientation of the integrated provirus",
             "DistTo5p"=paste("Distance to the DNA encoding 5 prime end of the",geneOfInterest,"transcript, in base pairs"), 
             "BreakSites"="Estimated abundance from number of independent linker addition sites (higher numbers indicate more abundant)", 
             "SiteRank"="Ranking of sites by abundance (1 indicates highest abundance)", 
             "SiteType"=paste("Nearest gene to integration site. For sites distant from",geneOfInterest,", this may not be",geneOfInterest,"itself."))
    
    kable(as.matrix(res), format="html", row.names=TRUE, 
          caption="Column descriptions")
  }
  
  ### geneOfInterest sites table ###   
  tableOfSites <- unique(sites.qc[rows,c("patient", "timepoint", "geneType",
                                         "celltype", "posID" ,"nrst5pOncoGeneDist", 
                                         "estAbundance1", "estAbundance1Rank")])
  tableOfSites <- arrange(tableOfSites, patient,timepoint,celltype,
                          abs(nrst5pOncoGeneDist), posID)
  names(tableOfSites) <- col.keys[names(tableOfSites)]
  
  ## curate timepoint ordering! ##
  tps <- sortTimePoints(tableOfSites$Tp)
  tableOfSites$Tp <- factor(tableOfSites$Tp, levels=names(tps))
  tableOfSites <- tableOfSites[order(tableOfSites$Tp),]
  
  tabcap <- paste(geneOfInterest,"sites")
  kable(tableOfSites, caption=tabcap, format="html", row.names=FALSE)       
} else {
  cat(paste("<strong>No",geneOfInterest,"sites found for in this patient.</strong>"))
}
```

## Integration sites near IKZF1

Due to recently developed interests, we extend the LMO2 analysis shown in previous section to include IKZF1. Results are summarized below.

```{r}
geneOfInterest <- "IKZF1"
printIt <- FALSE
```

```{r IKZF1, ref.label='geneOfInterest', results="asis", fig.width=10, fig.height=7}
```

## Integration sites near CCND2
  
We also wanted to investigate whether any sites where near CCND2 gene which has been a bad actor previously. Here are the results.
  
```{r}
geneOfInterest <- "CCND2"
printIt <- FALSE
```

```{r CCND2, ref.label='geneOfInterest', results="asis", fig.width=10, fig.height=7}
```

 
## Integration sites near HMGA2

We also wanted to investigate whether any sites where near HMGA2 gene which has been a bad actor previously. Here are the results.

```{r}
geneOfInterest <- "HMGA2"
printIt <- FALSE
```

```{r HMGA2, ref.label='geneOfInterest', results="asis", fig.width=10, fig.height=7}
```

## Integration sites near MECOM/EVI1

```{r}
geneOfInterest <- "MECOM|EVI1"
printIt <- FALSE
```

```{r MECOM, ref.label='geneOfInterest', results="asis", fig.width=10, fig.height=7}
```

## Cancer related genes & integration sites 
  
Out of all the aspects concerned in a gene therapy trial, integration near an oncogene or a cancer related gene is the most undesirable scenerio. The plot below compares the ratio of integration sites within 50kb of oncogene TSS to non-oncogenes. The number atop each bar indicates the number of sites within 50kb of oncogene TSS and the total number of unique sites is shown at the very top.

```{r onco_plots, fig.width=10, fig.height=8}
for(f in c("nrstOncoGeneDist","nrst5pOncoGeneDist")) {
  sites.qc[,f][is.na(sites.qc[,f])] <- Inf
}

sites.qc$oncoWithin50kb <- abs(sites.qc$nrstOncoGeneDist)<=50000 | 
                           abs(sites.qc$nrst5pOncoGeneDist)<=50000
sites.qc$geneWithin50kb <- abs(sites.qc$nrstGeneDist)<=50000 | 
                           abs(sites.qc$X5pnrstGeneDist)<=50000
sites.qc$onco5pWithin50kb <- abs(sites.qc$nrst5pOncoGeneDist)<=50000
sites.qc$gene5pWithin50kb <- abs(sites.qc$X5pnrstGeneDist)<=50000

test <- ddply(.data=sites.qc, .(timepoint,celltype), summarize,
              TotalSites=length(unique(posID)),
              onco50kb=length(unique(posID[oncoWithin50kb])),
              gene50kb=length(unique(posID[geneWithin50kb])),
              ratio50kb=onco50kb/gene50kb,
              
              onco5p50kb=length(unique(posID[onco5pWithin50kb])),
              gene5p50kb=length(unique(posID[gene5pWithin50kb])),
              ratio5p50kb=onco5p50kb/gene5p50kb,
              
              geneConOnco50kb=length(unique(posID[geneWithin50kb & oncoWithin50kb])),
              percCon50kb=geneConOnco50kb/gene50kb,
              
              geneConOnco5p50kb=length(unique(posID[gene5pWithin50kb & onco5pWithin50kb])),              
              percCon5p50kb=geneConOnco5p50kb/gene5p50kb)

test$timepoint <- with(test, factor(timepoint, 
                                    levels=names(sortTimePoints(unique(timepoint)))))
test$ratio5p50kb[is.na(test$ratio5p50kb)] <- 0
nrows <- ifelse(any(rowSums(xtabs(~celltype+timepoint,test))>6),2,1)
p <- qplot(data=test, x=timepoint, y=ratio5p50kb, geom='bar', stat='identity', 
           xlab="Timepoint") +
  geom_text(aes(y=(ratio5p50kb+0.007),label=onco5p50kb)) +
  geom_text(aes(y=(max(test$ratio5p50kb)+0.05), label=TotalSites)) +
  facet_grid(.~celltype, scales="free_x", space="free_x") +
  scale_y_continuous(name="Percent of Sites within 50kb of Oncogene TSS", 
                     labels=percent, expand=c(0,0.01)) +
  theme(axis.text.x=element_text(angle=45,hjust=1))
print(p)
```

## Do any clones account for greater than 20% of the total?
  
For some trials, a reporting criteria is whether any cell clones expand to account for greater than 20% of all clones. This is summarized below for subject `r patient`. Abundance is estimated using the SonicLength method. Data such as this must, of course, be interpreted in the context of results from other assays.

```{r TwentyPercSites, results="asis"}
### Sites >20% of data by Alias ###
rows <- sites.qc$estAbundance1Prop >= .2

if(any(rows)) {
  test2 <- arrange(unique(sites.qc[rows,c("patient", "timepoint", "celltype", "posID",
                                          "estAbundance1", "estAbundance1Prop",
                                          "estAbundance1Rank", "geneType")]),
                   patient,timepoint,celltype,plyr::desc(estAbundance1Prop),posID)
  tps <- sortTimePoints(test2$timepoint)  
  test2$timepoint <- factor(test2$timepoint, levels=names(tps))
  test2 <- arrange(test2, patient,timepoint,celltype,estAbundance1Rank)  
  names(test2) <- col.keys[names(test2)]
  test2$RelativeAbundance <- percent(test2$RelativeAbundance)
  kable(test2, caption="Sites >20% of the Total", format="html", row.names=FALSE)  
} else { 
  cat("<strong>No sites found in this patient which are greater than 20% of the total data.</strong>")        
}
```

### Do any multihit account for greater than 20% of the total?

Up until now, all the analysis has been looking at unique integration sites. But it is also helpful to look at reads finding multiple equally good scoring hits/places in the genome which can be reffered to as 'Multihits'. If an integration site occurred within a repeat element (i.e. Alus, LINE, SINE, etc), then it might be helpful to access those sites for potential detrimental effects. These collection of sequences are binned and analyzed separately due to their ambiguity. To make some sense of these multihits, we bin any sequence(s) which share 1 or more genomic locations hence forming psuedo-collections which can be reffered to as OTUs (operation taxonomic units). Once the OTUs are formed, we compare breakpoints of unique sites and multihits. The idea is to see if there are any multihits which higher in abundance than a unique site in a given sample. Below is a table similar to the one shown previously except we show any site which might be greater than 20\% of all clones in the data.

```{r Top10All, results="asis"}
rows <- sites.all$estAbundance1Prop >= .2

if(any(rows)) {
  toprint <- unique(subset(sites.all, estAbundance1Prop >= .2)
                    [,setdiff(names(sites.all), c('posID','Chr','strand','Position',
                                                 'Sequence','otuID','Alias',
                                                 'AliasOTUid','estAbundance1PropRank',
                                                 'timepointDay'))
                     ])
  toprint$Aliasposid <- NULL
  tps <- sortTimePoints(toprint$timepoint)  
  toprint$timepoint <- factor(toprint$timepoint, levels=names(tps))
  toprint <- arrange(toprint, patient,timepoint,celltype,estAbundance1Rank)	
  toprint$isMultiHit <- NULL
  names(toprint) <- col.keys[names(toprint)]
  toprint$RelativeAbundance <- percent(toprint$RelativeAbundance)
  kable(toprint, caption="All Sites >20% of the Total", 
        format="html", row.names=FALSE)  
} else { 
  cat("<strong>No sites found in this patient which are >20% of the total data after combining multihits.</strong>")  
}	
```
## Global View

The plot below shows percent of integration sites shared between any two samples. The overlap is performed by checking how many sites from 'Sample1' are within `r sanitize(paste(windows,collapse=","))`(bp) of sites in 'Sample2'. 

```{r overlaps, fig.width=10, fig.height=10}
Aliases <- unique(sites.qc$Alias)
sampleType.combos <- expand.grid(Aliases, Aliases, stringsAsFactors=FALSE)
names(sampleType.combos) <- paste0("Sample",1:2)

sites.gr <- with(unique(sites.qc[,c('Chr','strand','Alias','Position')]), 
                 GRanges(seqnames=Chr, strand=strand, Alias=Alias,
                         IRanges(start=Position, width=1)))

## add union of sites length for percent total at later stage ##
test.gr <- split(sites.gr, paste(mcols(sites.gr)$Alias))
sampleType.combos$UnionSites <- mcmapply(function(x,y) length(union(x,y)),
                                         test.gr[sampleType.combos[,1]],
                                         test.gr[sampleType.combos[,2]],
                                         mc.cores = detectCores())
rm(test.gr)

overlap.res <- lapply(windows, function(w) {
  overlap.res <- as.data.frame(findOverlaps(sites.gr, maxgap=w, 
                                            ignoreSelf=T, ignoreRedundant=FALSE))
  overlap.res$Sample1 <- mcols(sites.gr)$Alias[overlap.res$queryHits]
  overlap.res$Sample2 <- mcols(sites.gr)$Alias[overlap.res$subjectHits]
  overlap.res <- count(overlap.res,c("Sample1","Sample2"))  
  
  ## add samples which found no overlap and set frequency to 0 ##
  rows <- !with(sampleType.combos, 
                paste(Sample1,Sample2)) %in% with(overlap.res,
                                                  paste(Sample1,Sample2))
  if(any(rows))
    overlap.res <- rbind(overlap.res, cbind(sampleType.combos[rows,1:2],freq=0))
  overlap.res$WindowSize <- w
  overlap.res
})

overlap.res <- do.call(rbind, overlap.res)
names(overlap.res)[grepl("freq",names(overlap.res))] <- "TotalOverlap"

rm(sites.gr)
cleanit <- gc()  

overlap.res <- merge(overlap.res,sampleType.combos)
overlap.res$PercOverlap <- with(overlap.res,TotalOverlap/UnionSites)
if(any(overlap.res$PercOverlap>1)) {
  stop("Percent overlap data doesn't add up to 1! 
       Please check the input data for duplicates.")
}

## maintain order of timepoint...so make labels manually for the plot ##
possible.combos <- paste(names(timepoint),
                         rep(sort(celltype),each=length(timepoint)), sep=":")
overlap.res$Sample1 <- factor(sub("(.+):(.+)", "\\2-\\1", overlap.res$Sample1), 
                              levels=sub("(.+):(.+)", "\\2-\\1", 
                                         intersect(possible.combos, 
                                                   overlap.res$Sample1)))
overlap.res$Sample2 <- factor(sub("(.+):(.+)", "\\2-\\1", overlap.res$Sample2), 
                              levels=sub("(.+):(.+)", "\\2-\\1", 
                                         intersect(possible.combos, 
                                                   overlap.res$Sample2)))

overlap.res$PercOverlap[with(overlap.res,Sample1==Sample2)] <- 1
overlap.res$Bin <- cut(100*(overlap.res$PercOverlap),
                       breaks=100*(seq(0,1,by=.05)), include.lowest=T)

p <- qplot(data=overlap.res, Sample1, Sample2, geom="tile", fill=PercOverlap) + 
  facet_grid(.~WindowSize, labeller = label_both) + 
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1)) +
  scale_fill_gradientn(name="Overlap", label=percent, 
                       colours = rev(brewer.pal(4, "BuPu")))
print(p + ggtitle("Percent of Sites Overlapping between Samples"))
```

### Overlap Venn diagrams

An ideal way to depict overlap would be via venn diagrams, however, things become quickly complicated when number of samples go beyond four. Hence we break down comparisons in combinations of four.

```{r vennDiagrams, fig.height=5, fig.show='hide'}
tps <- with(unique(sites.qc[,c("posID","timepoint")]), split(posID,timepoint))
all_tps <- names(sortTimePoints(names(tps)))
first_tp <- all_tps[1]
last_tp <- all_tps[length(all_tps)]
if(length(tps)>1) {
  if(length(tps)>4) {
    combos <- combn(all_tps[!all_tps %in% c(first_tp,last_tp)], 2)
    if(ncol(combos)<15) {
      Vstem <- Venn(tps)
      for(f in 1:ncol(combos)) {
        plot(Vstem[,c(first_tp, last_tp, combos[,f])], 
             doWeights = FALSE, type='ellipses')
        }
      }  
    } else {
      if(length(tps)>3)
        plot(Venn(tps), doWeights = FALSE, type='ellipses')
      else
        plot(Venn(tps), doWeights = FALSE)
    }
}
```
```{r venn_Table, results='asis'}
plotFiles <- list.files(path="figureByPatient", pattern="vennDiagrams.*.png", 
                        full.names=TRUE)
if(length(plotFiles)>0) {
  plotFiles <- paste0("<img src='",plotFiles,"' alt='vennDiagrams'>")
  plotFiles <- matrix(plotFiles, ncol=2)
  dups <- names(which(table(plotFiles)>1))
  if(length(dups)>0) {
    plotFiles[tail(which(plotFiles==dups),1)] <- ""
    }
  kable(plotFiles, row.names=FALSE,format="html")
} else {
  cat("<strong>Not enough timepoint available to do this analysis.</strong>")  
}
```

### SiteTypes

The plot in previous section summarizes overlapping sites at the genomic coordinate level. However, integration sites are often represented by the gene they are in or nearby (SiteType). The plot below summarizes which 'SiteTypes' are often found to be abundant across samples relative to the entire landscape. The sites with abundance greater than 5% and rank within top two are colored.

```{r global_siteType, fig.width=10, fig.height=9}
sums <- aggregateSiteTypes(sites.qc, siteTypeVar="geneType")
sums$timepoint <- sub("(.+):(.+)","\\1",sums$Alias)
sums$celltype <- sub("(.+):(.+)","\\2",sums$Alias)
sums <- merge(sums, with(sums, getRanks(Props, siteType, Alias, "Ranks2")),
              by.x=c("siteType", "Alias"),by.y=c("posID","grouping"))
sums$SiteRank <- factor(with(sums, ifelse(Ranks2<4, Ranks2,">3")), levels=c(1:3,">3"))
sums$siteType2 <- with(sums, ifelse(Props>=0.05 & Ranks2<3, siteType, ""))

tps <- names(sortTimePoints(as.character(sums$timepoint)))
sums$timepoint <- factor(as.character(sums$timepoint), levels=tps)
counts <- count(sums, c("timepoint","celltype"))
sums$celltype <- factor(sums$celltype, levels=unique(counts$celltype))

# set custom color scale #
siteTypeCols <- structure(gg_color_hue(length(unique(sums$siteType2))),
                          names=unique(sums$siteType2))
siteTypeCols[names(siteTypeCols)==""] <- "grey70"

p <- qplot(data=sums, x=timepoint, y=Props, colour=siteType2, xlab="Timepoint",
           geom="jitter", position = position_jitter(h = 0.001)) +
  geom_hline(y=0.05,linetype='dotted') +
  scale_colour_manual(values=siteTypeCols) +
  scale_y_continuous(name="Relative Abundance", labels=percent, expand=c(0,0.01)) + 
  facet_grid(.~celltype, scales="free_x", space="free_x") + 
  theme(axis.text.x=element_text(angle=45, hjust=1, vjust=1))
p <- direct.label(p,"smart.grid")
print(p)
```

What is the most frequently occuring SiteType in subject `r patient`?

```{r wordle, fig.width=8, fig.height=8}
counts <- count(sums,"siteType"); names(counts)[1] <- "word"
suppressWarnings(plotWordCloud(counts, scale=c(3,0.5), min.freq=1, max.words=500, 
                               rot.per = 0, 
                               colors=c(colSets("Set1")[-6],colSets("Paired"))))
```

```{r overlaps_REvsFrag, results='hide', eval=FALSE}
### Overlap Analysis Restriction Enzyme Vs Fragmentase

Often it is of interest to investigate whether integration sites recovered using Restriction Enzyme(s) are seen again using Fragmentase method or not. In the analysis to follow, we divide each sample data by the isolation method and test how many integration sites overlap using window size of 5bp. For comparability, samples labelled with general celltypes such as WB/Blood were replaced with PBMC.

test <- sites.qc
test$isFrag <- grepl('FRAG',as.character(test$enzyme))

## change celltype WB/Blood to PBMC for comparability ##
test$Alias <- gsub("WB|Blood","PBMC",test$Alias,ignore.case=T)

toCheck <- pmin(xtabs(~Alias+isFrag,test),1)
toCheck <- toCheck[rowSums(toCheck)>1,]
if(is.null(dim(toCheck))) {
  cat("<strong>No samples found which used both isolation methods.</strong>")
} else {
  toCheck.aliases <- rownames(toCheck)
  sites.gr <- with(unique(test[test$Alias %in% toCheck.aliases,
                               c('Chr','strand','Alias','Position','isFrag')]), 
                   GRanges(seqnames=Chr, strand=strand, Alias=Alias, isFrag=isFrag, 
                           IRanges(start=Position, width=1)))
  
  
  test.gr <- split(sites.gr, paste(mcols(sites.gr)$Alias))
  
  overlap.res <- sapply(test.gr, findOverlaps, maxgap=5, 
                        ignoreSelf=T, ignoreRedundant=T)
  
  ## find overlap & get union of sites per isolation method for percent total ##
  overlap.res <- lapply(test.gr, 
                        function(x) {
                          res <- as.data.frame(findOverlaps(x, maxgap=5, 
                                                            ignoreSelf=TRUE, 
                                                            ignoreRedundant=TRUE))
                          res$isFrag1 <- mcols(x)$isFrag[res$queryHits]
                          res$isFrag2 <- mcols(x)$isFrag[res$subjectHits]
                          union.res <- length(union(subset(x,mcols(x)$isFrag),
                                                    subset(x,!mcols(x)$isFrag)))
                          cbind(Sample=as.character(x$Alias[1]),
                                count(res,c("isFrag1","isFrag2")),                                
                                UnionSites=union.res)
                        })
  
  rm("sites.gr","test.gr")
  cleanit <- gc()  
  
  overlap.res <- do.call(rbind, overlap.res)
  names(overlap.res)[grepl("freq",names(overlap.res))] <- "TotalOverlap"
  overlap.res$PercentOverlap <- percent(with(overlap.res,TotalOverlap/UnionSites))
  
  overlap.res$Tp <- sub("(.+):.+","\\1",overlap.res$Sample)
  overlap.res$Cell <- sub(".+:(.+)","\\1",overlap.res$Sample)
  tps <- sortTimePoints(overlap.res$Tp)  
  overlap.res$Tp <- factor(overlap.res$Tp, levels=names(tps))
  overlap.res <- arrange(overlap.res, Tp, Cell)
  wanted.cols <- c("Tp", "Cell", "TotalOverlap", "UnionSites", "PercentOverlap")

  kable(overlap.res[,wanted.cols], row.names=FALSE, format="html", digits = 0,
        caption="Sites Overlaping between Isolation Methods")
}
```
